File: client-scalaz/src/main/scala/org/openapitools/client/api/ConnectionGroupResponse.scala
package org.openapitools.client.api

import argonaut._
import argonaut.EncodeJson._
import argonaut.DecodeJson._

import org.http4s.{EntityDecoder, EntityEncoder}
import org.http4s.argonaut._
import org.joda.time.DateTime


import ConnectionGroupResponse._

case class ConnectionGroupResponse (
  valid: Option[Boolean],
message: Option[String],
version: Option[Double],
serializeNulls: Option[Boolean],
startTimeLog: Option[Long],
errorCode: Option[String],
request: Option[List[NameStringValueResponse]],
connectionGroupId: Option[Long],
name: Option[String],
groupType: Option[GroupType],
thumbnailURL: Option[String],
subGroups: Option[List[ConnectionGroupResponse]],
description: Option[String],
active: Option[Boolean],
connectionCount: Option[Integer],
connections: Option[List[ConnectionResponse]],
groupPermissions: Option[GroupPermissions],
ownerId: Option[Long],
returning: Option[String])

object ConnectionGroupResponse {
  import DateTimeCodecs._
  sealed trait GroupType
  case object `PRIVATE` extends GroupType
  case object FRIENDS extends GroupType
  case object TRUSTED extends GroupType
  case object BLOCKED extends GroupType
  case object FOLLOWERS extends GroupType
  case object FOLLOWING extends GroupType
  case object PENDING extends GroupType
  case object REQUEST extends GroupType

  object GroupType {
    def toGroupType(s: String): Option[GroupType] = s match {
      case "`PRIVATE`" => Some(`PRIVATE`)
      case "FRIENDS" => Some(FRIENDS)
      case "TRUSTED" => Some(TRUSTED)
      case "BLOCKED" => Some(BLOCKED)
      case "FOLLOWERS" => Some(FOLLOWERS)
      case "FOLLOWING" => Some(FOLLOWING)
      case "PENDING" => Some(PENDING)
      case "REQUEST" => Some(REQUEST)
      case _ => None
    }

    def fromGroupType(x: GroupType): String = x match {
      case `PRIVATE` => "`PRIVATE`"
      case FRIENDS => "FRIENDS"
      case TRUSTED => "TRUSTED"
      case BLOCKED => "BLOCKED"
      case FOLLOWERS => "FOLLOWERS"
      case FOLLOWING => "FOLLOWING"
      case PENDING => "PENDING"
      case REQUEST => "REQUEST"
    }
  }

  implicit val GroupTypeEnumEncoder: EncodeJson[GroupType] =
    EncodeJson[GroupType](is => StringEncodeJson(GroupType.fromGroupType(is)))

  implicit val GroupTypeEnumDecoder: DecodeJson[GroupType] =
    DecodeJson.optionDecoder[GroupType](n => n.string.flatMap(jStr => GroupType.toGroupType(jStr)), "GroupType failed to de-serialize")

  implicit val ConnectionGroupResponseCodecJson: CodecJson[ConnectionGroupResponse] = CodecJson.derive[ConnectionGroupResponse]
  implicit val ConnectionGroupResponseDecoder: EntityDecoder[ConnectionGroupResponse] = jsonOf[ConnectionGroupResponse]
  implicit val ConnectionGroupResponseEncoder: EntityEncoder[ConnectionGroupResponse] = jsonEncoderOf[ConnectionGroupResponse]
}

File: client-clojure/src/sirqul_io_t_platform/specs/connection_group_response.clj
(ns sirqul-io-t-platform.specs.connection-group-response
  (:require [clojure.spec.alpha :as s]
            [spec-tools.data-spec :as ds]
            [sirqul-io-t-platform.specs.name-string-value-response :refer :all]
            [sirqul-io-t-platform.specs.connection-group-response :refer :all]
            [sirqul-io-t-platform.specs.connection-response :refer :all]
            [sirqul-io-t-platform.specs.group-permissions :refer :all]
            )
  (:import (java.io File)))


(def connection-group-response-data
  {
   (ds/opt :valid) boolean?
   (ds/opt :message) string?
   (ds/opt :version) float?
   (ds/opt :serializeNulls) boolean?
   (ds/opt :startTimeLog) int?
   (ds/opt :errorCode) string?
   (ds/opt :request) (s/coll-of name-string-value-response-spec)
   (ds/opt :connectionGroupId) int?
   (ds/opt :name) string?
   (ds/opt :groupType) string?
   (ds/opt :thumbnailURL) string?
   (ds/opt :subGroups) (s/coll-of connection-group-response-spec)
   (ds/opt :description) string?
   (ds/opt :active) boolean?
   (ds/opt :connectionCount) int?
   (ds/opt :connections) (s/coll-of connection-response-spec)
   (ds/opt :groupPermissions) group-permissions-spec
   (ds/opt :ownerId) int?
   (ds/opt :returning) string?
   })

(def connection-group-response-spec
  (ds/spec
    {:name ::connection-group-response
     :spec connection-group-response-data}))
